pages/advanced-concepts/optimisations/cairo-profiler.md

```diff
@@ -0,0 +1,282 @@
+# Cairo Profiler Technical Guide: Gas Optimization
+
+**Understanding Available Metrics**
+
+The profiler collects several key metrics per function call:
+
+```cairo
+
+measurements = {
+    "steps": n_steps,
+    "memory_holes": n_memory_holes,
+    "calls": 1,  // count of function invocations
+    "l2_l1_message_sizes": message_sizes_sum,
+    // builtin counters (pedersen, range_check, etc.)
+}
+```
+
+## Running the Profiler
+
+```cairo
+cairo-profiler <path_to_trace_data> [OPTIONS]
+
+Options:
+  -o, --output-path <PATH>                    Output path [default: profile.pb.gz]
+  --show-details                              Show contract addresses and selectors
+  --max-function-stack-trace-depth <DEPTH>    Maximum depth of function tree [default: 100]
+  --split-generics                            Split generic functions by type
+  --show-inlined-functions                    Show inlined function calls
+  --versioned-constants-path <PATH>           Custom resource costs file
+
+```
+
+### Example: Optimizing Storage Access
+
+### Unoptimized Code
+
+```cairo
+#[starknet::contract]
+mod UnoptimizedContract {
+    #[storage]
+    struct Storage {
+        balances: LegacyMap<ContractAddress, u256>,
+        allowances: LegacyMap<(ContractAddress, ContractAddress), u256>,
+    }
+
+    #[external]
+    fn transfer_all_from_list(
+        ref self: ContractState,
+        from: ContractAddress,
+        to_list: Array<ContractAddress>
+    ) {
+        // Unoptimized: Multiple storage reads/writes
+        let len = to_list.len();
+        let mut i: usize = 0;
+        loop {
+            if i >= len {
+                break;
+            }
+            let recipient = *to_list.at(i);
+            let balance = self.balances.read(from);
+            self.balances.write(from, 0);
+            self.balances.write(recipient, balance);
+            i += 1;
+        };
```

Comment:
Could you move this in a `listing`

Reply:
Hey @julio4 do you mean the `listing` folder?

Reply:
Okay